Refactor

zbahati opened the pull request:
- Create input_validate to validate dates
- Create library_manager
- Create menu.rb
- Modify the app.rb file

# Finish

Commit:
Ft Add classes

- Add linters
- Add Gemfile
- Add Person, Student, Teachers classes

Finishes

Commit:
fix linters error

Commit:
Ft Decorator

Add classes Trim, Capitalize, decorator
Add inherits from decorator
Add decorator to inherits from Nameable class
Test the decorator

# Finishes

Commit:
ft setup -association

Add classes book, rental, classroom
Add relationship rentals to the person
Add relationship book, to the student
Add relationship of book to the rental and parent

Commit:
Fix linters error

Commit:
Correct all changes requested

Commit:
Add basic ui for the console application

Commit:
Fix linters error

Commit:
Adding quit method to exit the program

Commit:
Ft Refactor-code

Create input_validate to validate dates
Create library_manager
create menu.rb
Modify the app.rb file

Commit:
Fix rubocop errors

Commit:
fix error

Errytagedesign committed:
Add librarymanager

Errytagedesign committed:
Add createBook and rental

frank1738 commented on app.rb:
```diff
@@ -0,0 +1,24 @@
+require_relative 'student_class'
+require_relative 'teacher_class'
+require_relative 'classroom_class'
+require_relative 'book_class'
+require_relative 'rental_class'
+
+library_manager = LibraryManager.new
+menu = Menu.new(library_manager)
+
+loop do
+  menu.display_menu
+  choice = gets.chomp.to_i
+
+  if menu.valid_choice?(choice)
+    menu.handle_choice(choice)
+  else
+    puts 'Invalid choice. Please choose a valid option.'
+  end
+
+  break if choice == 8
+end
```
- [ ] 👋 Hi there, Great work on your recent PR! I appreciate your efforts in refactoring the code. It's a positive step towards maintaining a clean and organized codebase. 🧹✨. One important thing to note is that it's highly recommended to use `main.rb` as the entry point for your application. If you choose a different name, please make sure to document it in the readme so that users can easily find the file to run. 📝👀
- [ ] I did notice that all of the options were not working as expected and were throwing errors. When refactoring, it's crucial to ensure that the functionality is maintained

![image](https://github.com/zbahati/school_library_oop/assets/98453979/a81722bc-976c-4c81-9697-022ea8144504)

####  Pay special attention to input validation and error handling to prevent unexpected input issues.

frank1738 reviewed (changes requested):
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

<img src="https://media3.giphy.com/media/qLWdMYX1NYF2g/giphy-downsized-medium.gif"/>


## Highlights

✅ Some of the code was refactored and improved 👍 



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Commit:
Fix validating input

Commit:
Change the entry point of the system from app.rb to main.rb

Commit:
Add how to run the system by add the entry file as main.rb

Shedrack-Sunday commented on menu.rb:
```diff
@@ -0,0 +1,163 @@
+    end
+  end
+
+  def create_student
+    puts 'Enter student\'s name:'
+    name = gets.chomp
+    puts 'Enter student\'s age:'
+    age = gets.chomp
+
+    if InputValidator.validate_integer_input(age)
+      student = @library_manager.create_person(age.to_i, name)
+      puts "Student #{student.name} (ID: #{student.id}) created successfully."
+    else
+      puts 'Invalid age. Please enter a valid age.'
+    end
```
- [ ] From the video prompt provided, you're meant to first create a person, then ask if the user is going to be a student or teacher, and ask for parent permission when creating a person. 
- In your case, you have separate options for the requirement, and your input does not ask for parent permission. Kindly fix this issue 👍 

### Video prompt:
![image](https://github.com/zbahati/school_library_oop/assets/50990925/59e1f2ae-6a41-4578-b6fb-c562a39742f3)

### Your view:
![image](https://github.com/zbahati/school_library_oop/assets/50990925/bf060d7a-0812-469f-aae3-c10fdf1f6594)

Shedrack-Sunday reviewed (changes requested):
Hello @zbahati        👋 

# CHANGES REQUIRED ♻️

Impressive work on the project so far👏
You've done extremely well so far in this project,  However, there are some adjustments you need to make to move to the next project. 💯

![](https://user-images.githubusercontent.com/80612925/157397398-00904e26-60ec-46cd-9510-240661c250ef.gif)

## Highlights   
- Good use of GitHub features ✔️   
- Nice work with the Ruby features ✔️
- Good job with class calls   ✔️  

# - Required Changes  🌵  

- [x] Please make sure all class parameters are added appropriately  as it is in the video prompt


_Check the comment below




# Optional suggestions 

Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing really improves your application.



Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

Commit:
Ft refactor

Commit:
Ft refactor
Add two new option in person
Add teacher option
Add student option
on the student add check permision of parent

nedjwab reviewed (approved):
# 🟢 Project Approved! 🏆

## Hi  @zbahati                     👋

Your project is complete! ✔️✔  🏁

Your code was very professional 💯

There is nothing else to say other than... _it's time to merge it_ :shipit:
Congratulations! 🎉

 
## Highlights 🌟

- ✔️ Descriptive commmit messsages
- ✔️ Descriptive PR description
- ✔️ No linter Errors


- The code is top-notch! 💯 and have a huge potential.

Cheers and Happy coding!👏👏👏

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.


_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
